openghg/analyse/_scenario.py

```diff
@@ -0,0 +1,1453 @@
+                obs_data = obs_data.resample(indexer={"time": resample_period}, base=base).mean()
+
+            else:
+                resample_period = resample_to
+                footprint_data = footprint_data.resample(indexer={"time": resample_period}, base=base).mean()
+                obs_data = obs_data.resample(indexer={"time": resample_period}, base=base).mean()
+
+        return obs_data, footprint_data
+
+    def combine_obs_footprint(self,
+                              resample_to: str = "coarsest",
+                              platform: Optional[str] = None,
+                              cache: Optional[bool] = True,
+                              recalculate: Optional[bool] = False) -> Dataset:
+        """
+        Combine observation and footprint data so these are on the same time
+        axis. This will both slice and resample the data to align this axis.
+
+        - Data is slices to smallest timeframe spanned by both footprint and obs
+        - Data is resampled according to resample_to input and using the mean
+        - Data is combined into one dataset
+
+        Args:
+            resample_to: Resample option to use for averaging:
+                          - either one of ["coarsest", "obs", "footprint"] to match to the datasets
+                          - or using a valid pandas resample period e.g. "2H".
+                         Default = "coarsest".
+            platform: Observation platform used to decide whether to resample
+            cache: Cache this data after calculation. Default = True.
+
+        Returns:
+            xarray.Dataset: Combined dataset aligned along the time dimension
+
+            If cache is True:
+                This data will be also be cached as the ModelScenario.scenario attribute.
+        """
+
+        self._check_data_is_present(need=["obs", "footprint"])
+        obs = cast(ObsData, self.obs)
+        footprint = cast(FootprintData, self.footprint)
+
+        # Return any matching cached data
+        if self.scenario is not None and not recalculate:
+            if self.scenario.attrs["resample_to"] == resample_to:
+                return self.scenario
+
+        # As we're not processing any satellite data yet just set tolerance to None
+        tolerance = None
+        if platform is None:
+            platform = self._get_platform()
+
+        # Align and merge the observation and footprint Datasets
+        aligned_obs, aligned_footprint = self._align_obs_footprint(resample_to=resample_to, platform=platform)
+        combined_dataset = combine_datasets(dataset_A=aligned_obs, dataset_B=aligned_footprint, tolerance=tolerance)
+
+        # Transpose to keep time in the last dimension position in case it has been moved in resample
+        combined_dataset = combined_dataset.transpose(..., "time")
+
+        # Save the observation data units
+        try:
+            mf = obs.data["mf"]
+            units: Optional[float] = float(mf.attrs["units"])
```

Comment:
I removed most of the times I noticed Optional[], not being linked to a None as I understood mypy better - I can remove this one too.

Reply:
We should allow `units` to be None here as sometimes this will not be specified. I can update code around this to make it clear.

Reply:
In fact I had to do that because of the next part of the `try-except` statement - where units is set to None. This was to make sure `mypy` understood the typing here.

Reply:
OK, sounds sensible.